src/diagnostics.jl

```diff
@@ -55,25 +52,174 @@ function velocity_div!(grid::RegularCartesianGrid, u, v, w, div)
     end
 end
 
-function run_diagnostic(model::Model, diag::VelocityDivergenceChecker)
-    u, v, w = model.velocities.u.data, model.velocities.v.data, model.velocities.w.data
-    div = model.stepper_tmp.fC1
+@inline _prod(data, ::Nothing, i, j, k) = @inbounds data[i, j, k]
+@inline _prod(data1,    data2, i, j, k) = @inbounds data1[i, j, k] * data2[i, j, k]
+
+# Work-inefficient inclusive scan that reduces along the xy dimensions.
+# Useful for computing mean(; dim=[1, 2]) lol.
+# Modified from: https://github.com/JuliaGPU/CUDAnative.jl/blob/master/examples/scan.jl
+@hascuda function gpu_accumulate_xy!(Rxy::CuDeviceArray{T}, Rx, data1, data2, op::Function) where T

```

Comment:
Well, it's meant to be temporary until the functionality is available in CuArrays.jl. See this open issue: https://github.com/JuliaGPU/CuArrays.jl/issues/68

It's a very general function (you could accumulate multiplications, or any binary operation), but I adapted it to accumulate in the x,y dimensions to give us horizontal averages.

So right now I've butchered it so that it's easy to use with the diagnostics I defined, although it would be easy to generalize it.

Reply:
> As for now, I would like to use output rather than diagnostics --- so the functionality I need is a function that computes the horizontal average of a 3D field and stores in a preallocated 1D field (or array --- either one).

On that note, this is exactly what `gpu_accumulate_xy!` does. Although it computes the sum so you have to normalize by `Nx*Ny` at the end.

Reply:
Maybe a docstring would be helpful to explain what it does?

Reply:
Ok, I think I am starting to understand what is being done here. 

A few comments:

* it's probably not necessary to append the name of this function with `gpu_`
* why do we dispatch on the first argument? Is there another version of `gpu_accumulate` that works with `CuArray` (or something?)
* a better name might be `accumulate_binary_op`
* is the output `Rxy`? Maybe just call it `output`?
* I think it'd be good to also define a function `horizontal_sum` for which `op=+`.

Reply:
> On that note, this is exactly what gpu_accumulate_xy! does.

It seems like `accumulate` takes two input fields though (data1 and data2). Wouldn't a horizontal sum/average just take one field as input?